### PR TITLE
[67] Novelty dashboard charts

### LINKED ISSUE #67: Charts for novelty measures
Continues the work from #54 and #64. 

Here we will write code to generate altair charts that could be used for the final project dashboard. 

These might include:

- Charts showing most novel topics, for a given year and given dataset (research vs patents)
- Chart showing document counts versus topic novelty - to evaluate popularity versus novelty of topics
- Chart characterising temporal dynamics of topic novelty (eg time series and/or magnitude vs growth)

Note that we might want to allow user to choose between our two different novelty measurement approaches (document vs topic-pair based novelty)

Perhaps we would also like to show the top-most (or least) novel papers/patents for a chosen topic?

We also might want to add charts showing the biggest differences in novelty between research and patents.

### LINKED ISSUE #73: Use "publication_date" for patent novelty measures
fyi @emily-bicks, will update asap

### PR DESCRIPTION
**Pull request by beingkk**
Closes #67
Closes #73 
---

# Description

The main contributions of this PR are:
- `utils/novelty_charts.py`: New utils for producing novelty charts
- `novelty_chart_examples.py`: New notebook with output tables/charts and text descriptions 

Hope this helps in implementing the charts in the streamlit dashboard! 

I've tried to refactor as much as possible of the chart and table production to the `novelty_charts.py` 

I've also added text descriptions describing the methodology and interpretation of the charts. Hope they are clear, but please feel free to adjust.

**NB:** Unfortunately I'm not here for the remainder of this week and next week (taking leave till first week of April). Of course happy to make adjustments to the PR when I come back in April, but I appreciate you'll want to move faster and merge it already. I hope it's OK for you to make any essential changes yourself? Thank you!

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

**Comment by beingkk** [comment]
Just FYI, I have refactored the code and added also time series charts.

I will add text descriptions in the notebook tomorrow morning, which will help explain what each chart is showing.

Hope this will be useful!

File: dap_aria_mapping/utils/novelty_charts.py [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+        topic_to_document_dict (Dict[str: List[str]]): Dictionary mapping topics to documents
+        topic_names (Dict[str: str]): Dictionary mapping topic labels to topic names
+        most_novel (bool, optional): Whether to get the most novel documents or the least novel documents.
+        topic (str, optional): Topic for which to get the mostl/least novel documents. If None, get the most novel documents for all topics.
+        year (int, optional): Year for which to get the most/lesat novel documents.
+        top_n (int, optional): Number of documents to return.
+        min_topics (int, optional): Minimum number of topics a document
+            must be associated with to be included in the results.
+        output_columns (list, optional): Columns to show in the output.
+    """
```
Comment (ampudia19) [inline dap_aria_mapping/utils/novelty_charts.py:99]: Missing return docstring.

File: dap_aria_mapping/utils/novelty_charts.py [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+    return (end_novelty - start_novelty) / start_novelty
+
+
+def calculate_magnitude_growth(
+    topic_novelty_df: pd.DataFrame,
+    topic_names: Dict[str, str],
+    start_year: int = 2018,
+    end_year: int = 2022,
+) -> pd.DataFrame:
+    """_summary_
```
Comment (ampudia19) [inline dap_aria_mapping/utils/novelty_charts.py]: Missing docstring description.

File: dap_aria_mapping/utils/novelty_charts.py [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+def chart_top_topics_bubble(data: pd.DataFrame, values_label: str = "Topic novelty"):
+    """
+    Chart top topics as a bubble chart
+
+    Args:
+        data (pd.DataFrame): Dataframe with top topics with the
+            following columns: topic_name, novelty_score, doc_counts
+
+    Returns:
+        Altair chart
```
Comment (ampudia19) [inline dap_aria_mapping/utils/novelty_charts.py]: alt.Chart

File: dap_aria_mapping/utils/novelty_charts.py [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+    """
+    Chart showing topic novelty over time
+
+    Args:
+        topic_novelty_df (pd.DataFrame): Topic novelty dataframe with columns
+            topic_name, topic_doc_novelty, doc_counts
+        topics_to_show (List[str]): List of topics  (labels) show in the char
+        values_label (str, optional): Label for the y-axis
+        show_legend (bool, optional): Whether to show the legend
+    """
```
Comment (ampudia19) [inline dap_aria_mapping/utils/novelty_charts.py:483]: missing return docstring

File: dap_aria_mapping/utils/novelty_utils.py [hunk 5]
```diff
@@ -358,3 +358,20 @@ def pair_to_topic_novelty(
...
         )
     )
+
+def create_topic_to_document_dict(
+    document_novelty_df,
+    document_df,
+    id_column="work_id",
+    document_title_column="display_name"
+):
+    """Generates dictionary for faster lookups"""
```
Comment (ampudia19) [inline dap_aria_mapping/utils/novelty_utils.py]: Missing hints and incomplete docstring.

**Review by ampudia19 — APPROVED** [review]
Good stuff, only minor adjustments to docstrings but otherwise pipelines look good.

**Pull request merged**